Update pagination service from invalid

RedNesto opened the pull request:
Includes a couple of fixes.

From a couple of simple tests it seems to work correctly.

Closes #2990 #2991 #2992 #2993 #2994 #2995 #2996

RedNesto committed:
Make common.service.pagination compile

Also use namespaced commands to avoid ambiguity with mods/plugins commands

RedNesto committed:
Couple of fixes

Make sure we compile with UTF-8 or it breaks special characters used in the pagination list
Attempt to work around a Style#getBold stackoverflow
Fix footer padding being 1 characters too long (not a fan of decrementing but I found no better way)
Fix padding color always using title's color
Remove unused methods and parameters

RedNesto committed:
Register pagination command on startup

dualspiral commented on src/main/java/org/spongepowered/common/command/manager/SpongeCommandManager.java:
```diff
@@ -524,6 +526,15 @@ public void init() {
+        try {
+            SpongeParameterizedCommandRegistrar.INSTANCE.register(
+                    Launcher.getInstance().getCommonPlugin(),
+                    ((SpongePaginationService) Sponge.getServiceProvider().paginationService()).createPaginationCommand(),
```
This should only run if the service wasn't replaced, this will throw a class cast exception otherwise.

RedNesto replied:
Oh right, I assumed it can't be replaced for some reason.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -143,105 +131,94 @@ private SourcePaginations getPaginationStateForPlayer(Player source, boolean cre
-        protected ActivePaginationCommandElement(@Nullable Text key) {
-            super(key);
-        }
+    private class ActivePaginationParser implements ValueParser<ActivePagination> {
```
Any reason that this is split into `ValueParser` and `ValueCompleter`? In general, they should just be one `ValueParameter` if this is the case ([the docs say if a `ValueParameter` is passed it will use the completer there](https://github.com/SpongePowered/SpongeAPI/blob/api-8/src/main/java/org/spongepowered/api/command/parameter/Parameter.java#L979-L983)).

RedNesto replied:
I missed that part, will do !

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -143,105 +131,94 @@ private SourcePaginations getPaginationStateForPlayer(Player source, boolean cre
-                .executor(childDispatcher)
-                .description(t("Helper command for paginations occurring"))
+        return Command.builder()
+                .parameters(paginationIdParameter, firstOf(pageParameter,
```
All static calls must be qualified - no static imports for this.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -143,105 +131,94 @@ private SourcePaginations getPaginationStateForPlayer(Player source, boolean cre
                 }).build();
 
-        CommandElement pageArgs = integer(t("page"));
+        Parameter.Value<Integer> pageParameter = integerNumber().setKey("page").build();
```
@Zidane likes final things. 

(When possible, variables and classes should be final)

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java:
```diff
@@ -28,23 +28,27 @@
-    private Text footer;
-    private Text paginationSpacer = Text.of("=");
+    private Component footer;
+    private Component paginationSpacer = EQUALS_TEXT;
```
Static acesses from the same class should also be qualified with the class name.

dualspiral reviewed (changes requested):
Mostly style, but a potential crash here too.

RedNesto committed:
Address feedback

- merge ActivePaginationParser & Completer
- remove static imports except for Preconditions
- fix possible CCE when registering pagination command

RedNesto committed:
Add finals everywhere

RedNesto committed:
Replace anonymous guava Function with method reference

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java:
```diff
@@ -218,82 +213,57 @@ int getWidth(Component text) {
-    private Component finalizeBuilder(Component text,TextComponent.Builder build) {
-        return build.style(text.style()).build();
+    private Component finalizeBuilder(final Component text, final TextComponent.Builder build) {
+        return build.style(text.style().toBuilder().build()).build();
```
Is there a reason for this `toBuilder().build()`? The following compiles fine:

```
final Component component = TextComponent.empty();
TextComponent.builder().style(component.style()).build();
```

I haven't run it though - if there is a problem that you've worked around, we need to report it to Adventure.

RedNesto replied:
Oh I didn't put a comment there to explain.
It seems to be needed to avoid StackOverflowExceptions that happen later for some reason. I mentioned it to kashike and AFAIK we're not sure why this happens.

dualspiral replied:
Is there an issue reported for this? I imagine it's ~~probably on the Adventure end~~ but it needs to be reported somewhere. (edit: it wasn't)

Once done, add a comment linking to that issue so if some poor soul wants more information (and perhaps wants to have a stab at fixing it), they can do so.

RedNesto replied:
Well I don't even know if the issue lies in the library or the SpongeCommon glue code. I should talk to kashike about it again I think.

dualspiral replied:
Oh, I just looked and I know exactly what's going on. It's basically the fact that the Minecraft `Style` object is being cached for a given Adventure `Style` object, but when you use Minecraft's `Style#setStyle` it also sets the parent of the child's style. In this case, the parent and the child have been supplied with the same Style reference, so when the parent style is set... it's itself.

Sponge error. We basically should not be caching mutable objects (edit: without copying them when we use them).

dualspiral replied:
https://github.com/SpongePowered/SpongeCommon/commit/44be61c0dbc79069f82c174b63d7a68aa254c1b7

RedNesto replied:
Amazing, thank you ! I hope this is the last force push I have to do for this PR.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java:
```diff
@@ -306,12 +276,12 @@ private Component finalizeBuilder(Component text,TextComponent.Builder build) {
                 .append(text)
                 .append(spaces)
-                .style(text.style())
+                .style(text.style().toBuilder().build())
```
As above.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationList.java:
```diff
@@ -46,19 +45,23 @@
-    private final Optional<Text> footer;
-    private final Text paginationSpacer;
+    private final Iterable<Component> contents;
+    private final Optional<Component> title;
```
More of a note for the future - but storing an optional just feels _wrong_. It's being used to avoid optional (re)creation in getters, but... how often are the getters called? Not often I'm sure! The `sendTo` method actually unwraps the optional into a null.

Perhaps this is a good time to ditch the optionals and just have nullable fields.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -0,0 +1,216 @@
+                .setExecutor(pageExecutor)
+                .build();
+
+        //We create the child manually in order to force that paginationElement is required for all children + fallback
```
Comment should be ditched.

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -0,0 +1,216 @@
+
+        @Override
+        public List<String> complete(final CommandContext context) {
+            return context.getCause().first(Audience.class)
```
`.first(Audience.class)` -> `.getAudience()` (audiences do have special handling in command causes)

dualspiral commented on src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java:
```diff
@@ -0,0 +1,216 @@
+        public Optional<? extends ActivePagination> getValue(final Parameter.Key<? super ActivePagination> parameterKey,
+                final ArgumentReader.Mutable reader,
+                final CommandContext.Builder context) throws ArgumentParseException {
+            final Audience source = context.getCause().first(Audience.class)
```
`.first(Audience.class)` -> `.getAudience()`

RedNesto committed:
Address feedback - pass 2

- stop copying Styles in PaginationCalculator to avoid StackOverflows,
 fixed in 44be61c0dbc79069f82c174b63d7a68aa254c1b7 thanks dual !
- replace Optional fields by nullable ones
- remove outdated comment
- use CommandCause#getAudience instead of Cause#first

Also fix a possible NPE in ActivePaginationParameter#complete
